Analysis/source/CalcMulticlassDNN.cxx

```diff
@@ -0,0 +1,376 @@
+  std::vector<std::string> ulong64Names = { "evt" };
+  std::vector<std::string> floatNames = { "b1_pt", "b1_eta", "b1_phi", "b1_m", "b1_DeepFlavour",
+    "b1_DeepFlavour_CvsB", "b1_DeepFlavour_CvsL", "b1_HHbtag", "b2_pt", "b2_eta", "b2_phi", "b2_m",
+    "b2_DeepFlavour", "b2_DeepFlavour_CvsB", "b2_DeepFlavour_CvsL", "b2_HHbtag", "VBF1_pt",
+    "VBF1_eta", "VBF1_phi", "VBF1_m", "VBF1_DeepFlavour", "VBF1_DeepFlavour_CvsB",
+    "VBF1_DeepFlavour_CvsL", "VBF1_HHbtag", "VBF2_pt", "VBF2_eta", "VBF2_phi", "VBF2_m",
+    "VBF2_DeepFlavour", "VBF2_DeepFlavour_CvsB", "VBF2_DeepFlavour_CvsL", "VBF2_HHbtag", "tau1_pt",
+    "tau1_eta", "tau1_phi", "tau1_m", "tau2_pt", "tau2_eta", "tau2_phi", "tau2_m", "MET_pt",
+    "MET_phi" };
+
+  // register them in input maps and set branch addresses
+  for (const auto& name : boolNames) {
+    boolInputs_.emplace(name, 0.);
+    inTree->SetBranchAddress(name.c_str(), &boolInputs_.at(name));
+  }
+  for (const auto& name : intNames) {
+    intInputs_.emplace(name, 0.);
+    inTree->SetBranchAddress(name.c_str(), &intInputs_.at(name));
+  }
+  for (const auto& name : ulong64Names) {
+    ulong64Inputs_.emplace(name, 0);
+    inTree->SetBranchAddress(name.c_str(), &ulong64Inputs_.at(name));
+  }
+  for (const auto& name : floatNames) {
+    floatInputs_.emplace(name, 0.);
+    inTree->SetBranchAddress(name.c_str(), &floatInputs_.at(name));
+  }
+}
+
+void FeatureProvider::calculate() {
+  // check if objects are set
+  bool b1Set = intInputs_.at("b1_valid") == 1;
+  bool b2Set = intInputs_.at("b2_valid") == 1;
+  bool vbfj1Set = intInputs_.at("VBF1_valid") == 1;
+  bool vbfj2Set = intInputs_.at("VBF2_valid") == 1;
+  bool lep1Set = floatInputs_.at("tau1_pt") > 0;
+  bool lep2Set = floatInputs_.at("tau2_pt") > 0;
+  bool bHSet = b1Set && b2Set;
+  bool tauHSet = lep1Set && lep2Set;
+  bool vbfjjSet = vbfj1Set && vbfj2Set;
+
+  // define vectors for objects
+  TLorentzVector b1, b2, vbfj1, vbfj2, lep1, lep2, bH, tauH, vbfjj;
+  b1.SetPtEtaPhiM(floatInputs_.at("b1_pt"), floatInputs_.at("b1_eta"), floatInputs_.at("b1_phi"),
+      floatInputs_.at("b1_m"));
+  b2.SetPtEtaPhiM(floatInputs_.at("b2_pt"), floatInputs_.at("b2_eta"), floatInputs_.at("b2_phi"),
+      floatInputs_.at("b2_m"));
+  vbfj1.SetPtEtaPhiM(floatInputs_.at("VBF1_pt"), floatInputs_.at("VBF1_eta"),
+      floatInputs_.at("VBF1_phi"), floatInputs_.at("VBF1_m"));
+  vbfj2.SetPtEtaPhiM(floatInputs_.at("VBF2_pt"), floatInputs_.at("VBF2_eta"),
+      floatInputs_.at("VBF2_phi"), floatInputs_.at("VBF2_m"));
+  lep1.SetPtEtaPhiM(floatInputs_.at("tau1_pt"), floatInputs_.at("tau1_eta"),
+      floatInputs_.at("tau1_phi"), floatInputs_.at("tau1_m"));
+  lep2.SetPtEtaPhiM(floatInputs_.at("tau2_pt"), floatInputs_.at("tau2_eta"),
+      floatInputs_.at("tau2_phi"), floatInputs_.at("tau2_m"));
+  bH = b1 + b2;
+  tauH = lep1 + lep2;
+  vbfjj = vbfj1 + vbfj2;
+
+  // loop through features and set values
+  for (auto& it : features_) {
```

Comment:
To improve the readability of the code I propose to set features in a different way.
```c++
size_t set_counter = 0;
const auto setFeature = [&](const std::string& name, float value, bool valid) {
    auto iter = features_.find(name);
    if(iter != features_.end()) {
        iter->second = valid ? value : hmc::features::EMPTY;
        ++set_counter;
    }
};

setFeature("bjet1_pt", event.b1_pt, event.b1_valid);
...

if(set_counter != features_.size())
    throw exception("Some features were not set"); // can be made more verbose.
```